File: sorts/countingsort/counting_sort.go
/**
 * Copyright © https://github.com/microwind All rights reserved.
 * @author: jarryli@gmail.com
 * @version: 1.0
 */
package main

import (
  "fmt"
  "time"
)

/**
 * 计数排序标准版
 */
func countingSort1(arr []int) []int {
  // 计算最大值与最小值
  var max = arr[0]
  var min = arr[0]
  var arrLen = len(arr)
  for i := 1; i < arrLen; i++ {
    if arr[i] > max {
      max = arr[i]
    }
    if arr[i] < min {
      min = arr[i]
    }
  }

  // 计数数组的长度在最大和最小差值+1
  var countList = make([]int, max-min+1)
  for i := 0; i < arrLen; i++ {
    // 把下标减去min值，以便减少计数数组的长度，最小项的坐标为0，以此递增
    var idx = arr[i] - min
    // 根据待排序项给对应下标的位置标记加1
    countList[idx] += 1
  }

  var idx = 0
  // 遍历计数数组，覆盖原数组
  for i := 0; i < len(countList); i++ {
    // 下标若大于0，则取出来，如果相同则取多次
    val := countList[i]
    for j := 0; j < val; j++ {
      if val > 0 {
        // 取出的值要+min还原
        arr[idx] = i + min
        idx++
      }
    }
  }

  return arr
}

// 计数排序2
func countingSort2(arr []int) []int {
  var max = arr[0]
  var min = arr[0]
  arrLen := len(arr)
  for i := 1; i < arrLen; i++ {
    if arr[i] > max {
      max = arr[i]
    }
    if arr[i] < min {
      min = arr[i]
    }
  }
  var output = make([]int, arrLen)
  var countList = make([]int, max-min+1)

  // 按每个数组项出现次数计数，从最小项的下标为0
  for _, item := range arr {
    // 当前前减去最小项得到下标，最后取出数据要还原
    var idx = item - min
    countList[idx]++
  }

  // fmt.Println("countList after count: ", countList)
  // 把上一项的值个数存入到当前项，相当于排序
  for i := 1; i < len(countList); i++ {
    countList[i] += countList[i-1]
  }
  // fmt.Println("countList after sort: ", countList)

  // 按位置还原数据，下一个索引记录了上一个的坐标值
  for i := 0; i < arrLen; i++ {
    // 当前项来自原始数组减去-min
    item := arr[i] - min
    // 根据当前项从计数数组里找到目标位置
    idx := countList[item] - 1
    // 输出数据加上min进行还原
    output[idx] = item + min
    // 取出一项计数则减少一个
    countList[item] -= 1

    fmt.Println("[item, idx, countList[idx], output] :", item, idx, countList[idx], output)
  }

  return output
}

// test
func main() {
  fmt.Println("counting sort1:")
  time1 := time.Now()
  data1 := []int{20, 11, 0, -10, 9, 9, 6, 30, 11, 15, 13, 80}
  fmt.Println(countingSort1(data1[:]))
  fmt.Println("sort1 end. cost:", time.Since(time1))

  fmt.Println("counting sort2:")
  time2 := time.Now()
  data2 := []int{20, 11, 0, -10, 9, 9, 6, 30, 11, 15, 13, 80}
  fmt.Println(countingSort2(data2[:]))
  fmt.Println("sort2 end. cost:", time.Since(time2))
}

/*
jarry@jarrys-MacBook-Pro countingsort % go version
go version go1.19.5 darwin/amd64
jarry@jarrys-MacBook-Pro countingsort % go run counting_sort.go
counting sort1:
[-10 0 6 9 9 11 11 13 15 20 30 80]
sort1 end. cost: 28.526µs
counting sort2:
[item, idx, countList[idx], output] : 30 9 1 [0 0 0 0 0 0 0 0 0 20 0 0]
[item, idx, countList[idx], output] : 21 6 1 [0 0 0 0 0 0 11 0 0 20 0 0]
[item, idx, countList[idx], output] : 10 1 1 [0 0 0 0 0 0 11 0 0 20 0 0]
[item, idx, countList[idx], output] : 0 0 0 [-10 0 0 0 0 0 11 0 0 20 0 0]
[item, idx, countList[idx], output] : 19 4 1 [-10 0 0 0 9 0 11 0 0 20 0 0]
[item, idx, countList[idx], output] : 19 3 1 [-10 0 0 9 9 0 11 0 0 20 0 0]
[item, idx, countList[idx], output] : 16 2 1 [-10 0 6 9 9 0 11 0 0 20 0 0]
[item, idx, countList[idx], output] : 40 10 1 [-10 0 6 9 9 0 11 0 0 20 30 0]
[item, idx, countList[idx], output] : 21 5 1 [-10 0 6 9 9 11 11 0 0 20 30 0]
[item, idx, countList[idx], output] : 25 8 1 [-10 0 6 9 9 11 11 0 15 20 30 0]
[item, idx, countList[idx], output] : 23 7 1 [-10 0 6 9 9 11 11 13 15 20 30 0]
[item, idx, countList[idx], output] : 90 11 2 [-10 0 6 9 9 11 11 13 15 20 30 80]
[-10 0 6 9 9 11 11 13 15 20 30 80]
sort2 end. cost: 137.672µs

==过程模拟==
data2 := []int{2, 1, 5, 3, 2}
fmt.Println(countingSort2(data2[:]))

原始 arr=[2,1,5,3,2]

1、填充计数：
min = 1
countingList=[0,0,0,0,0]
countingList[arr[i]-min] += 1
计数统计：countingList=[1, 2, 1, 0, 1]

2、计数排序：
计数统计：countingList[i] += countingList[i-1]
计数排序：countingList=[1, 3, 4, 4, 5]

3、按位置取出：
loop arr:

0:
    arr[i] = 2
    item = arr[i]-min = 1
    idx = countingList[item] - 1 = 2
    output[idx] = item + min = 2
    countingList[item] -= 1

    output = [0, 0, 2, 0, 0]
    countingList=[1, 2, 4, 4, 5]

1:
    arr[i] = 1
    item = arr[i]-min = 0
    idx = countingList[item] - 1 = 0
    output[idx] = item + min = 1
    countingList[item] -= 1

    output = [1, 0, 2, 0, 0]
    countingList=[0, 2, 4, 4, 5]

2:
    arr[i] = 5
    item = arr[i]-min = 4
    idx = countingList[item] - 1 = 4
    output[idx] = item + min = 5
    countingList[item] -= 1

    output = [1, 0, 2, 0, 5]
    countingList=[0, 2, 4, 4, 4]

3:
    arr[i] = 3
    item = arr[i]-min = 2
    idx = countingList[item] - 1 = 3
    output[idx] = item + min = 3
    countingList[item] -= 1

    output = [1, 0, 2, 3, 5]
    countingList=[0, 2, 3, 4, 4]

4:
    arr[i] = 2
    item = arr[i]-min = 1
    idx = countingList[item] - 1 = 1
    output[idx] = item + min = 2
    countingList[item] -= 1

    output = [1, 2, 2, 3, 5]
    countingList=[0, 1, 3, 4, 4]

*/

File: sorting/countingsort/counting_sort.dart
/**
 * Copyright © https://github.com/microwind All rights reserved.
 * 
 * @author: jarryli@gmail.com
 * @version: 1.0
 */

/**
 * 计数排序算法实现
 * 提供四种不同的实现方式，适合不同场景和性能需求
 */

import 'dart:io';

/**
 * 打印数组内容的辅助函数
 */
void printArray(List<int> arr, String label) {
  print('$label: [${arr.join(', ')}]');
}

/**
 * 性能测试辅助函数
 */
void performanceTest(Function(List<int>) sortFunc, List<int> arr, String name) {
  // 创建数组副本，避免修改原数组
  List<int> testArr = List.from(arr);
  printArray(testArr, '${name}原始数组');
  
  // 开始计时
  Stopwatch stopwatch = Stopwatch()..start();
  sortFunc(testArr);
  stopwatch.stop();
  
  print('$name: ${stopwatch.elapsedMilliseconds.toDouble() / 1000.0}');
  printArray(testArr, '${name}排序结果');
  print(''); // 空行分隔
}

// ==================== 主程序：算法演示和性能测试 ====================

// 测试数据：包含负数和重复元素的典型数组
const List<int> testData = [20, 11, 0, -10, 9, 9, 6, 30, 11, 15, 13, 80];

/**
 * 计数排序基础版本 - 标准实现
 * 
 * 算法原理：
 * 1. 统计每个元素出现的次数
 * 2. 计算累计计数以确定元素位置
 * 3. 根据累计计数将元素放到正确位置
 * 4. 从后向前遍历保证稳定性
 * 
 * 生活类比：就像统计班级学生成绩分布，
 * 先统计每个分数段有多少人，然后按分数段排序
 * 
 * 时间复杂度：O(n + k)，n为元素个数，k为数据范围
 * 空间复杂度：O(k) - 需要计数数组
 * 稳定性：稳定 - 从后向前遍历保持相等元素的相对位置
 */
void countingSort1(List<int> arr) {
  print('countingSort1 standard:');
  
  if (arr.isEmpty) {
    return;
  }
  
  // 找到最小值和最大值
  int min = arr.reduce((a, b) => a < b ? a : b);
  int max = arr.reduce((a, b) => a > b ? a : b);
  int range = max - min + 1;
  
  // 创建计数数组
  List<int> count = List.filled(range, 0);
  
  // 统计每个元素出现的次数
  for (int i = 0; i < arr.length; i++) {
    count[arr[i] - min]++;
  }
  
  // 计算累计计数
  for (int i = 1; i < count.length; i++) {
    count[i] += count[i - 1];
  }
  
  // 创建输出数组
  List<int> output = List.filled(arr.length, 0);
  
  // 从后向前遍历，保证稳定性
  for (int i = arr.length - 1; i >= 0; i--) {
    int value = arr[i];
    int index = value - min;
    output[count[index] - 1] = value;
    count[index]--;
  }
  
  // 复制回原数组
  for (int i = 0; i < arr.length; i++) {
    arr[i] = output[i];
  }
  
  printArray(arr, '排序后数组');
}

/**
 * 计数排序优化版本 - 内存优化
 * 
 * 算法思路：
 * 直接在原数组上进行修改，减少内存使用
 * 使用原地排序技术
 * 
 * 优化效果：
 * - 减少额外空间使用
 * - 更好的缓存性能
 * 
 * 时间复杂度：O(n + k)
 * 空间复杂度：O(k)
 * 稳定性：不稳定 - 原地修改可能影响稳定性
 */
void countingSort2(List<int> arr) {
  print('countingSort2 memory optimized:');
  
  if (arr.isEmpty) {
    return;
  }
  
  // 找到最小值和最大值
  int min = arr.reduce((a, b) => a < b ? a : b);
  int max = arr.reduce((a, b) => a > b ? a : b);
  int range = max - min + 1;
  
  // 创建计数数组
  List<int> count = List.filled(range, 0);
  
  // 统计每个元素出现的次数
  for (int value in arr) {
    count[value - min]++;
  }
  
  // 直接根据计数重构数组
  int index = 0;
  for (int i = 0; i < count.length; i++) {
    int remaining = count[i];
    while (remaining > 0) {
      arr[index] = i + min;
      index++;
      remaining--;
    }
  }
  
  printArray(arr, '排序后数组');
}

/**
 * 计数排序 - 负数单独处理版本
 * 
 * 算法思路：
 * 专门处理包含负数的情况
 * 分别处理负数和正数部分
 * 
 * 优化效果：
 * - 支持负数排序
 * - 保持算法稳定性
 * 
 * 时间复杂度：O(n + k)
 * 空间复杂度：O(k)
 * 稳定性：稳定
 */
void countingSort3(List<int> arr) {
  print('countingSort3 negative numbers:');
  
  if (arr.isEmpty) {
    return;
  }
  
  // 分离正数和负数
  List<int> negatives = [];
  List<int> positives = [];
  
  for (int value in arr) {
    if (value < 0) {
      negatives.add(value);
    } else {
      positives.add(value);
    }
  }
  
  // 排序负数部分（转换为绝对值排序后反转）
  if (negatives.isNotEmpty) {
    List<int> absNegatives = negatives.map((x) => -x).toList();
    
    // 使用独立的计数排序逻辑
    if (absNegatives.isNotEmpty) {
      // 找到最小值和最大值
      int min = absNegatives.reduce((a, b) => a < b ? a : b);
      int max = absNegatives.reduce((a, b) => a > b ? a : b);
      int range = max - min + 1;
      
      // 创建计数数组
      List<int> count = List.filled(range, 0);
      
      // 统计每个元素出现的次数
      for (int i = 0; i < absNegatives.length; i++) {
        count[absNegatives[i] - min]++;
      }
      
      // 计算累计计数
      for (int i = 1; i < count.length; i++) {
        count[i] += count[i - 1];
      }
      
      // 创建输出数组
      List<int> output = List.filled(absNegatives.length, 0);
      
      // 从后向前遍历，保证稳定性
      for (int i = absNegatives.length - 1; i >= 0; i--) {
        int value = absNegatives[i];
        int index = value - min;
        output[count[index] - 1] = value;
        count[index]--;
      }
      
      // 反转并恢复负号
      for (int i = 0; i < output.length; i++) {
        negatives[i] = -output[output.length - 1 - i];
      }
    }
  }
  
  // 排序正数部分
  if (positives.isNotEmpty) {
    // 使用独立的计数排序逻辑
    if (positives.isNotEmpty) {
      // 找到最小值和最大值
      int min = positives.reduce((a, b) => a < b ? a : b);
      int max = positives.reduce((a, b) => a > b ? a : b);
      int range = max - min + 1;
      
      // 创建计数数组
      List<int> count = List.filled(range, 0);
      
      // 统计每个元素出现的次数
      for (int i = 0; i < positives.length; i++) {
        count[positives[i] - min]++;
      }
      
      // 计算累计计数
      for (int i = 1; i < count.length; i++) {
        count[i] += count[i - 1];
      }
      
      // 创建输出数组
      List<int> output = List.filled(positives.length, 0);
      
      // 从后向前遍历，保证稳定性
      for (int i = positives.length - 1; i >= 0; i--) {
        int value = positives[i];
        int index = value - min;
        output[count[index] - 1] = value;
        count[index]--;
      }
      
      // 复制回正数数组
      for (int i = 0; i < output.length; i++) {
        positives[i] = output[i];
      }
    }
  }
  
  // 合并结果
  List<int> result = [];
  result.addAll(negatives);
  result.addAll(positives);
  
  // 复制回原数组
  for (int i = 0; i < arr.length; i++) {
    arr[i] = result[i];
  }
  
  printArray(arr, '排序后数组');
}

/**
 * 计数排序 - 桶优化版本
 * 
 * 算法思路：
 * 使用桶的概念，将相近的数值分组
 * 减少计数数组的大小
 * 适合大数据范围但分布稀疏的情况
 * 
 * 优化效果：
 * - 减少内存使用
 * 提高处理稀疏数据的效率
 * 
 * 时间复杂度：O(n + k)
 * 空间复杂度：O(k)
 * 稳定性：稳定
 */
void countingSort4(List<int> arr) {
  print('countingSort4 bucket optimized:');
  
  if (arr.isEmpty) {
    return;
  }
  
  // 找到最小值和最大值
  int min = arr.reduce((a, b) => a < b ? a : b);
  int max = arr.reduce((a, b) => a > b ? a : b);
  int range = max - min + 1;
  
  // 确定桶的大小（可以调整以优化性能）
  int bucketSize = (range / 10).clamp(1, double.infinity).toInt();
  int bucketCount = ((range + bucketSize - 1) / bucketSize).clamp(5, double.infinity).toInt();
  
  // 创建桶
  List<List<int>> buckets = List.generate(bucketCount, (_) => []);
  
  // 将元素分配到桶中
  for (int value in arr) {
    int bucketIndex;
    if (value == max) {
      bucketIndex = bucketCount - 1;
    } else {
      bucketIndex = (value - min) ~/ bucketSize;
    }
    buckets[bucketIndex].add(value);
  }
  
  // 对每个桶进行排序并合并结果
  List<int> result = [];
  
  for (int i = 0; i < bucketCount; i++) {
    if (buckets[i].isNotEmpty) {
      // 使用独立的计数排序逻辑
      List<int> bucketArray = List.from(buckets[i]);
      
      if (bucketArray.isNotEmpty) {
        // 找到最小值和最大值
        int min = bucketArray.reduce((a, b) => a < b ? a : b);
        int max = bucketArray.reduce((a, b) => a > b ? a : b);
        int range = max - min + 1;
        
        // 创建计数数组
        List<int> count = List.filled(range, 0);
        
        // 统计每个元素出现的次数
        for (int j = 0; j < bucketArray.length; j++) {
          count[bucketArray[j] - min]++;
        }
        
        // 计算累计计数
        for (int j = 1; j < count.length; j++) {
          count[j] += count[j - 1];
        }
        
        // 创建输出数组
        List<int> output = List.filled(bucketArray.length, 0);
        
        // 从后向前遍历，保证稳定性
        for (int j = bucketArray.length - 1; j >= 0; j--) {
          int value = bucketArray[j];
          int index = value - min;
          output[count[index] - 1] = value;
          count[index]--;
        }
        
        // 复制回桶数组
        for (int j = 0; j < output.length; j++) {
          bucketArray[j] = output[j];
        }
      }
      
      // 将排序后的桶内容复制到结果数组
      result.addAll(bucketArray);
    }
  }
  
  // 复制回原数组
  for (int i = 0; i < arr.length; i++) {
    arr[i] = result[i];
  }
  
  printArray(arr, '排序后数组');
}

// ==================== 算法测试和性能对比 ====================

void main() {
  // 测试1：标准版本
  performanceTest(countingSort1, testData, '标准版本');

  // 测试2：内存优化版本
  performanceTest(countingSort2, testData, '内存优化版本');

  // 测试3：负数处理版本
  performanceTest(countingSort3, testData, '负数处理版本');

  // 测试4：桶优化版本
  performanceTest(countingSort4, testData, '桶优化版本');

  print('=== 算法对比总结 ===');
  print('1. 标准版本：经典实现，稳定排序');
  print('2. 内存优化版本：原地修改，节省空间');
  print('3. 负数处理版本：支持负数，功能完整');
  print('4. 桶优化版本：分组处理，适合稀疏数据');
}

/*
打印结果
jarry@Mac heapsort % dart heap_sort.dart    
heapSort1 max-heap原始数组: 7, 11, 9, 10, 12, 13, 8
heapSort1 max-heap:
oringal array:
              7(0)
          /         \
        11(1)        9(2)
       /    \      /     \
   10(3)  12(4)  13(5)  8(6)

父节点步骤
1. maxHeapify: idx= 2 left= 5 right= 6 max= 5 size= 7

              7(0)
          /         \
        11(1)        13(2)
       /    \      /     \
   10(3)  12(4)  9(5)  8(6)

3. maxHeapify: idx= 1 left= 3 right= 4 max= 4 size= 7

              7(0)
          /         \
        12(1)        13(2)
       /    \      /     \
   10(3)  11(4)  9(5)  8(6)

5. maxHeapify: idx= 0 left= 1 right= 2 max= 2 size= 7

              13(0)
          /         \
        12(1)        7(2)
       /    \      /     \
   10(3)  11(4)  9(5)  8(6)

5.1 maxHeapify: idx= 5 left= 11 right= 12 max= 5 size= 7

              13(0)
          /         \
        12(1)        9(2)
       /    \      /     \
   10(3)  11(4)  7(5)  8(6)

子节点堆排序:
7. swap, child = 6:

              8(0)
          /         \
        12(1)        9(2)
       /    \      /     \
   10(3)  11(4)  7(5)  13(6)

7.1 maxHeapify: idx= 1 left= 3 right= 4 max= 4 size= 6

              12(0)
          /         \
        8(1)        9(2)
       /    \      /     \
   10(3)  11(4)  7(5)  13(6)

7.1 maxHeapify: idx= 4 left= 9 right= 10 max= 4 size= 6

              12(0)
          /         \
        11(1)        9(2)
       /    \      /     \
   10(3)  8(4)  7(5)  13(6)

8. swap, child = 5:

              7(0)
          /         \
        11(1)        9(2)
       /    \      /     \
   10(3)  8(4)  12(5)  13(6)

8.1 maxHeapify: idx= 1 left= 3 right= 4 max= 3 size= 5

              11(0)
          /         \
        7(1)        9(2)
       /    \      /     \
   10(3)  8(4)  12(5)  13(6)

8.1 maxHeapify: idx= 3 left= 7 right= 8 max= 3 size= 5

              11(0)
          /         \
        10(1)        9(2)
       /    \      /     \
   7(3)  8(4)  12(5)  13(6)

9. swap, child = 4:

              8(0)
          /         \
        10(1)        9(2)
       /    \      /     \
   7(3)  11(4)  12(5)  13(6)

9.1 maxHeapify: idx= 1 left= 3 right= 4 max= 1 size= 4

              10(0)
          /         \
        8(1)        9(2)
       /    \      /     \
   7(3)  11(4)  12(5)  13(6)

10. swap, child = 3:

              7(0)
          /         \
        8(1)        9(2)
       /    \      /     \
   10(3)  11(4)  12(5)  13(6)

10.1 maxHeapify: idx= 2 left= 5 right= 6 max= 2 size= 3

              9(0)
          /         \
        8(1)        7(2)
       /    \      /     \
   10(3)  11(4)  12(5)  13(6)

11. swap, child = 2:

              7(0)
          /         \
        8(1)        9(2)
       /    \      /     \
   10(3)  11(4)  12(5)  13(6)

11.1 maxHeapify: idx= 1 left= 3 right= 4 max= 1 size= 2

              8(0)
          /         \
        7(1)        9(2)
       /    \      /     \
   10(3)  11(4)  12(5)  13(6)

12. swap, child = 1:

              7(0)
          /         \
        8(1)        9(2)
       /    \      /     \
   10(3)  11(4)  12(5)  13(6)

排序后数组: 7, 8, 9, 10, 11, 12, 13
heapSort1 max-heap: 0.000ms
heapSort1 max-heap排序结果: 7, 8, 9, 10, 11, 12, 13

heapSort2 min-heap原始数组: 7, 11, 9, 10, 12, 13, 8
heapSort2 min-heap:
排序后数组: 7, 8, 9, 10, 11, 12, 13
heapSort2 min-heap: 0.000ms
heapSort2 min-heap排序结果: 7, 8, 9, 10, 11, 12, 13

heapSort3 iterative原始数组: 7, 11, 9, 10, 12, 13, 8
heapSort3 iterative:
排序后数组: 7, 8, 9, 10, 11, 12, 13
heapSort3 iterative: 0.000ms
heapSort3 iterative排序结果: 7, 8, 9, 10, 11, 12, 13

heapSort4 bottom-up原始数组: 7, 11, 9, 10, 12, 13, 8
heapSort4 bottom-up:
排序后数组: 7, 8, 9, 10, 11, 12, 13
heapSort4 bottom-up: 0.000ms
heapSort4 bottom-up排序结果: 7, 8, 9, 10, 11, 12, 13

heapSort5 non-recursive原始数组: 7, 11, 9, 10, 12, 13, 8
heapSort5 non-recursive:
current= 9  idx= 2  child= 5  size= 7
parent sort: 2 
: 7, 11, 13, 10, 12, 9, 8
current= 11  idx= 1  child= 3  size= 7
parent sort: 1 
: 7, 12, 13, 10, 11, 9, 8
current= 7  idx= 0  child= 1  size= 7
current= 7  idx= 2  child= 5  size= 7
parent sort: 0 
: 13, 12, 9, 10, 11, 7, 8
child start: parent=-1  child=6
current= 8  idx= 0  child= 1  size= 6
current= 8  idx= 1  child= 3  size= 6
child sort: 6 
: 12, 11, 9, 10, 8, 7, 13
current= 7  idx= 0  child= 1  size= 5
current= 7  idx= 1  child= 3  size= 5
child sort: 5 
: 11, 10, 9, 7, 8, 12, 13
current= 8  idx= 0  child= 1  size= 4
current= 8  idx= 1  child= 3  size= 4
child sort: 4 
: 10, 8, 9, 7, 11, 12, 13
current= 7  idx= 0  child= 1  size= 3
child sort: 3 
: 9, 8, 7, 10, 11, 12, 13
current= 7  idx= 0  child= 1  size= 2
child sort: 2 
: 8, 7, 9, 10, 11, 12, 13
child sort: 1 
: 7, 8, 9, 10, 11, 12, 13
排序后数组: 7, 8, 9, 10, 11, 12, 13
heapSort5 non-recursive: 0.000ms
heapSort5 non-recursive排序结果: 7, 8, 9, 10, 11, 12, 13

=== 算法对比总结 ===
1. heapSort1 max-heap：经典实现，升序排序
2. heapSort2 min-heap：反向思维，直观易懂
3. heapSort3 iterative：避免递归，性能优化
4. heapSort4 bottom-up：高效构建，减少比较
5. heapSort5 non-recursive：非递归堆化，父子节点计算函数
*/
